tests/eas/misfit.py

```diff
@@ -0,0 +1,400 @@
+            'tag' : 'misfit',
+            'flags' : ['ftrace', 'freeze_userspace'],
+        }
+
+        if 'cpufreq' in test_env.target.modules:
+            available_govs = test_env.target.cpufreq.list_governors(0)
+            conf['cpufreq'] = {'governor' : 'performance'}
+
+        return {
+            'wloads' : cls._get_wload(test_env),
+            'confs' : [conf],
+        }
+
+class StaggeredFinishes(_MisfitMigrationBase):
+    """
+    Test Misfit task migration happens at idle balance (staggered test case)
+
+    This test spawns nr_cpus 100% tasks. The tasks running on bigger-capacity
+    CPUs will finish first, and it is expected of them to instantly pull the
+    tasks running on smaller-capacity CPUs via idle-balance.
+    """
+
+    # How long the tasks will be pinned to their "starting" CPU. Doesn't have
+    # to be long (we just have to ensure they spawn there), so arbitrary value
+    pin_delay = 0.001
+
+    # Somewhat arbitrary delay - long enough to ensure
+    # rq->avg_idle > sysctl_sched_migration_cost
+    idling_delay = 1
+
+    # How long do we allow misfit tasks to stay on LITTLE CPUs when big CPUs
+    # are free
+    allowed_misfit_pct = 10
+
+    @classmethod
+    def _get_wload(cls, test_env):
+        # rt-app json file is assembled by hand because we need per-phase
+        # CPU-affinity, and that's not (yet) feasible in LISA
+        wload = {
+            "global": {
+                "calibration": "__PVALUE__",
+                "default_policy": "SCHED_OTHER",
+                "duration": -1,
+                "logdir" : "__LOGDIR__"
+            },
+            "tasks": {}
+        }
+
+        task_template = {
+            "delay": int(cls.idling_delay * 1000000),
+            "loop": 1,
+            "phases": OrderedDict([("pinned", {}), ("free", {})]),
+            "policy": "SCHED_OTHER"
+        }
+
+        for key in task_template["phases"].iterkeys():
+            task_template["phases"][key]["loop"] = 1
+
+        task_template["phases"]["pinned"]["run"] = int(1e6 * cls.pin_delay)
+        # We're pinning stuff in the first phase, so give it ample time to
+        # clean the pinned logic out of balance_interval
+        task_template["phases"]["free"]["run"] = int(
+            1e6 * 1.1 * cls._get_max_lb_interval(test_env))
+
+        for cpu in xrange(test_env.platform['cpus_count']):
+            task = deepcopy(task_template)
+            task["phases"]["pinned"]["cpus"] = [cpu]
+            wload["tasks"]["misfit_{}".format(cpu)] = task
+
+        wload_path = os.path.join(test_env.res_dir, 'staggered_template.json')
+        with open(wload_path, 'w') as fh:
+            res = json.dumps(wload)
+            # Tokens that represent integers musn't be escaped by quotes.
+            # The reason strings are used is to dump the json more easily.
+            for key in ['__PVALUE__']:
+                res = res.replace('"{}"'.format(key), key)
+            fh.write(res)
```

Comment:
I'm not really a big fan of this way to generate rt-app JSONs... especially now that we support cpumasks per phases since 0e5d5853d

Not testes, but I think something like this should do its job...

```python
pinned_time = # whatever you like
free_time = # whatever you like
all_cpus = range(test_env.platform['cpus_count']);
tasks = {}
for cpu in all_cpus:
    task_pinned = Periodic(duty_cycle_pct=100, duration_s=pinned_time, period_ms=16, cpu=cpu)
    task_free = Periodic(duty_cycle_pct=100, duration_s=free_time, period_ms=16, cpu=all_cpus)
    task = task_pinned + task_free
    tasks["task_{}".format(cpu)] = task.get()

rtapp.conf(
    kind='profile',
    params=tasks
)
```

Is the new API not good enough for your use-case?

Reply:
That new API wasn't there when I wrote this :p